lib/libzfs/libzfs_changelist.c

```diff
@@ -627,10 +632,16 @@ changelist_gather(zfs_handle_t *zhp, zfs_prop_t prop, int gather_flags,
 		}
 	}
 
-	clp->cl_pool = uu_avl_pool_create("changelist_pool",
-	    sizeof (prop_changenode_t),
-	    offsetof(prop_changenode_t, cn_treenode),
-	    compare_mountpoints, 0);
+	if (legacy)
+		clp->cl_pool = uu_avl_pool_create("changelist_pool",
+		    sizeof (prop_changenode_t),
+		    offsetof(prop_changenode_t, cn_treenode),
+		    compare_dataset_names, 0);
+	else
+		clp->cl_pool = uu_avl_pool_create("changelist_pool",
+		    sizeof (prop_changenode_t),
+		    offsetof(prop_changenode_t, cn_treenode),
+		    compare_mountpoints, 0);
```

Comment:
nit: how about writing it this way instead, it makes it clear that the only difference is the passed function.

```c
	clp->cl_pool = uu_avl_pool_create("changelist_pool",
￼ 	    sizeof (prop_changenode_t),
￼ 	    offsetof(prop_changenode_t, cn_treenode),
￼ 	    legacy ? compare_dataset_names : compare_mountpoints, 0);
```